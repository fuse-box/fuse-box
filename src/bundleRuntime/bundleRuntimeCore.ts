import { ITarget } from '../config/ITarget';
import { readFile } from '../utils/utils';

export const BUNDLE_RUNTIME_NAMES = {
  ARG_REQUIRE_FUNCTION: '__fusereq',
  BUNDLE_FUNCTION: 'bundle',
  GLOBAL_OBJ: '__fuse',
  INTEROP_REQUIRE_DEFAULT_FUNCTION: 'dt',
  REQUIRE_FUNCTION: 'r',
};

export type ICodeSplittingMap = {
  b: Record<number, { p: string }>;
};

export interface IBundleRuntimeCore {
  codeSplittingMap?: ICodeSplittingMap;
  interopRequireDefault?: boolean;
  isIsolated?: boolean;
  target: ITarget;
  typescriptHelpersPath?: string;
}

const INTEROP_DEFAULT = BUNDLE_RUNTIME_NAMES.INTEROP_REQUIRE_DEFAULT_FUNCTION;

function getCodeSplittingFunction(target: ITarget) {
  if (target === 'browser') {
    return `function lb(id, conf) {
  return new Promise(function(resolve, reject) {
    if (!conf.fns) {
      conf.fns = [resolve];
      var script = document.createElement('script');
      document.head.appendChild(script);
      script.onload = function() {
        if (modules[id]) {
          conf.fns.map(x => x(f.r(id)));
        } else reject('Resolve error of module ' + id + ' at path ' + conf.p);
        conf.fns = void 0;
      };
      script.src = conf.p;
    } else conf.fns.push(resolve);
  });
}`;
  }
  if (target === 'server') {
    return `function lb(id, conf) {
return new Promise(function(resolve, reject) {
  require(require('path').resolve(__dirname, conf.p));
if (modules[id]) {
  return resolve(f.r(id));
} else reject('Resolve error of module ' + id + ' at path ' + conf.p);
});
};`;
  }
  return '';
}

export function bundleRuntimeCore(props: IBundleRuntimeCore) {
  const { target } = props;
  const isIsolated = props.target === 'web-worker' || props.isIsolated;
  let coreVariable;
  // web-worker cannot share anything with bundle, hence it gets an isolated mode
  if (isIsolated) {
    coreVariable = `var f = {};`;
  } else if (target === 'browser' || target === 'electron') {
    // storing directly to browser window object
    coreVariable = `var f = window.${BUNDLE_RUNTIME_NAMES.GLOBAL_OBJ} = window.${BUNDLE_RUNTIME_NAMES.GLOBAL_OBJ} || {};`;
  } else if (target === 'server') {
    // storing to global object
    coreVariable = `var f = global.${BUNDLE_RUNTIME_NAMES.GLOBAL_OBJ} = global.${BUNDLE_RUNTIME_NAMES.GLOBAL_OBJ} || {};`;
  }

  let optional = '';
  if (props.interopRequireDefault) {
    optional += `f.${INTEROP_DEFAULT} = function (x) { return x !== undefined && x.default !== undefined ? x.default : x; };\n`;
  }
  if (props.codeSplittingMap) {
    optional += `\nvar cs = ${JSON.stringify(props.codeSplittingMap)};`;
    optional += '\n' + getCodeSplittingFunction(props.target);
  }

  let CODE = `${isIsolated ? `var ${BUNDLE_RUNTIME_NAMES.GLOBAL_OBJ} = ` : ''}(function() {
  ${coreVariable}
  var modules = f.modules = f.modules || {}; ${optional}
  f.${BUNDLE_RUNTIME_NAMES.BUNDLE_FUNCTION} = function(collection, fn) {
    for (var num in collection) {
      modules[num] = collection[num];
    }
    fn ? fn() : void 0;
  };
  f.c = {};
  f.${BUNDLE_RUNTIME_NAMES.REQUIRE_FUNCTION} = function(id) {
    var cached = f.c[id];
    if (cached) return cached.m.exports;
    var module = modules[id];
    if (!module) {
      ${props.codeSplittingMap ? 'var s = cs.b[id]; if (s) return lb(id, s);' : ''}
      throw new Error('Module ' + id + ' was not found');
    }
    cached = f.c[id] = {};
    cached.exports = {};
    cached.m = { exports: cached.exports };
    module(f.${BUNDLE_RUNTIME_NAMES.REQUIRE_FUNCTION}, cached.exports, cached.m);
    return cached.m.exports;
  }; ${isIsolated ? '\n\treturn f;' : ''}
})();`;

  if (props.typescriptHelpersPath) {
    const tsHelperContents = readFile(props.typescriptHelpersPath);
    CODE += '\n' + tsHelperContents;
  }

  return CODE;
}
