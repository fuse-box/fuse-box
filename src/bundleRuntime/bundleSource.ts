import { ITarget } from '../config/PrivateConfig';
import { IModule } from '../moduleResolver/Module';
import { Concat } from '../utils/utils';
import { BUNDLE_RUNTIME_NAMES, bundleRuntimeCore, IBundleRuntimeCore } from './bundleRuntimeCore';

export interface IBundleSourceProps {
  core?: IBundleRuntimeCore;
  isIsolated?: boolean;
  target: ITarget;

  withSourcemaps?: boolean;
}

export type IBundleSource = ReturnType<typeof bundleSource>;

const FuseName = BUNDLE_RUNTIME_NAMES.GLOBAL_OBJ;
const BundleFN = FuseName + '.' + BUNDLE_RUNTIME_NAMES.BUNDLE_FUNCTION;
const ReqFn = FuseName + '.' + BUNDLE_RUNTIME_NAMES.REQUIRE_FUNCTION;

export function bundleSource(props: IBundleSourceProps) {
  const isIsolated = props.target === 'web-worker' || props.isIsolated;

  const modules: Array<IModule> = [];
  let entries: Array<IModule> = [];
  let injection: string;
  let expose: Array<{ name: string; moduleId: number }>;
  const scope = {
    entries,
    expose,
    // user injection
    // for example inject some code after the bundle is ready
    injection,
    modules,
    generate: () => {
      const concat = new Concat(props.withSourcemaps, '', '\n');
      // start the wrapper for the entire bundle if required
      if (isIsolated) concat.add(null, `(function(){`);

      // adding core api if required
      if (props.core) concat.add(null, bundleRuntimeCore(props.core));

      concat.add(null, BundleFN + '({');

      let index = 0;
      const totalAmount = scope.modules.length;
      while (index < totalAmount) {
        const module = scope.modules[index];
        const isLast = index + 1 === totalAmount;
        if (module.contents) {
          if (module.publicPath) {
            concat.add(null, `\n// module-${module.id} ${module.publicPath}`);
          }
          concat.add(null, module.id + `: function(${BUNDLE_RUNTIME_NAMES.ARG_REQUIRE_FUNCTION}, exports, module){`);
          concat.add(
            props.withSourcemaps ? module.publicPath : null,
            module.contents,
            props.withSourcemaps ? module.sourceMap : undefined,
          );
          concat.add(null, '}' + (isLast ? '' : ','));
        }
        index++;
      }

      let injectionCode = [];
      // add entries
      // e.g __fuse.r(1)
      if (scope.entries) {
        for (const entry of scope.entries) {
          injectionCode.push(ReqFn + '(' + entry.id + ')');
        }
      }

      // add exposed variables
      // for example on nodejs that will be "exports" on browser "window"
      if (scope.expose) {
        let exposedGlobal;
        if (props.target === 'browser' || props.target === 'electron') exposedGlobal = 'window';
        else if (props.target === 'server') exposedGlobal = 'exports';
        // we cannot expose on web-worker target
        if (exposedGlobal) {
          for (const item of scope.expose) {
            injectionCode.push(`${exposedGlobal}[${JSON.stringify(item.name)}] = ${ReqFn + '(' + item.moduleId + ')'}`);
          }
        }
      }

      let readyFunction = '';
      if (injectionCode.length) readyFunction = `, function(){\n` + injectionCode.join('\n') + '\n}';

      concat.add(null, '}' + readyFunction + ')');
      // end the isolation
      if (isIsolated) concat.add(null, `})()`);
      return concat;
    },
  };
  return scope;
}
