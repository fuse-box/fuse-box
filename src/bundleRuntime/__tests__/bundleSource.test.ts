import { BUNDLE_RUNTIME_NAMES, bundleRuntimeCore, IBundleRuntimeCore } from '../bundleRuntimeCore';
import { bundleSource } from '../bundleSource';

const Fuse = BUNDLE_RUNTIME_NAMES.GLOBAL_OBJ;
const RequireFunc = BUNDLE_RUNTIME_NAMES.REQUIRE_FUNCTION;

function createInteropCall(args) {
  return Fuse + '.' + BUNDLE_RUNTIME_NAMES.INTEROP_REQUIRE_DEFAULT_FUNCTION + '(' + JSON.stringify(args) + ');';
}
describe('Bundle source test', () => {
  function runWithoutApi(code: string, props?: IBundleRuntimeCore): any {
    props = props || { target: 'browser' };

    const globalObj = props.target === 'browser' || props.target === 'electron' ? 'window' : 'global';
    const serverExports = {};
    const obj = {};
    const api = bundleRuntimeCore(props);
    code = api + `\n var ${Fuse} = ${globalObj}.${Fuse};\n` + code;

    var res = new Function(globalObj, 'exports', code);
    res(obj, props.target === 'server' ? serverExports : undefined);
    return {
      obj,
      serverExports,
      req: (num: number) => {
        return obj[Fuse][RequireFunc](num);
      },
    };
  }

  describe('Generic require', () => {
    it('should execute 1 file', () => {
      const source = bundleSource({ target: 'browser' });
      source.modules = [{ contents: 'exports.Foo = "bar"', id: 1 }];
      const result = source.generate();
      const { req } = runWithoutApi(result.content.toString());
      expect(req(1)).toEqual({ Foo: 'bar' });
    });

    it('should execute first file having 2 files', () => {
      const source = bundleSource({ target: 'browser' });
      source.modules = [
        { contents: 'exports.Foo = "bar"', id: 1 },
        { contents: 'exports.Second = true', id: 2 },
      ];
      const result = source.generate();
      const { req } = runWithoutApi(result.content.toString());
      expect(req(1)).toEqual({ Foo: 'bar' });
      expect(req(2)).toEqual({ Second: true });
    });

    it('should execute 1 file only once', () => {
      const source = bundleSource({ target: 'browser' });
      source.modules = [
        {
          contents: `
            let counter = 0;
            exports.counter = ++counter`,
          id: 1,
        },
      ];
      const result = source.generate();
      const { req } = runWithoutApi(result.content.toString());
      expect(req(1)).toEqual({ counter: 1 });
      expect(req(1)).toEqual({ counter: 1 });
    });
  });

  describe('Bundle entries', () => {
    it('should call entry 1', () => {
      const source = bundleSource({ target: 'browser' });

      source.modules = [
        {
          contents: `window.wasCalled = true;`,
          id: 1,
        },
      ];
      source.entries = [{ id: 1 }];
      const result = source.generate();
      const { obj } = runWithoutApi(result.content.toString());
      expect(obj.wasCalled).toEqual(true);
    });

    it('should call 2 entries', () => {
      const source = bundleSource({ target: 'browser' });

      source.modules = [
        {
          contents: `window.FirstCalled = true;`,
          id: 1,
        },
        {
          contents: `window.SecondCalled = true;`,
          id: 2,
        },
      ];
      source.entries = [{ id: 1 }, { id: 2 }];
      const result = source.generate();
      const { obj } = runWithoutApi(result.content.toString());
      expect(obj.FirstCalled).toEqual(true);
      expect(obj.SecondCalled).toEqual(true);
    });

    it('should polyfil default', () => {
      const source = bundleSource({ target: 'browser' });

      source.modules = [
        {
          contents: `window.FirstCalled = ${createInteropCall({ foo: 'bar' })}`,
          id: 1,
        },
      ];
      source.entries = [{ id: 1 }];
      const result = source.generate();
      const { obj } = runWithoutApi(result.content.toString(), { interopRequireDefault: true, target: 'browser' });
      expect(obj.FirstCalled).toEqual({ foo: 'bar' });
    });

    it('should take the existing default (obj)', () => {
      const source = bundleSource({ target: 'browser' });

      source.modules = [
        {
          contents: `window.FirstCalled = ${createInteropCall({ default: 'bar' })}`,
          id: 1,
        },
      ];
      source.entries = [{ id: 1 }];
      const result = source.generate();
      const { obj } = runWithoutApi(result.content.toString(), { interopRequireDefault: true, target: 'browser' });
      expect(obj.FirstCalled).toEqual('bar');
    });

    it('should take  default (number)', () => {
      const source = bundleSource({ target: 'browser' });

      source.modules = [
        {
          contents: `window.FirstCalled = ${createInteropCall(1)}`,
          id: 1,
        },
      ];
      source.entries = [{ id: 1 }];
      const result = source.generate();
      const { obj } = runWithoutApi(result.content.toString(), { interopRequireDefault: true, target: 'browser' });
      expect(obj.FirstCalled).toEqual(1);
    });
  });

  describe('Bundle expose', () => {
    it('should expose on window with target browser', () => {
      const source = bundleSource({ target: 'browser' });

      source.modules = [
        {
          contents: `exports.Foo = "bar"`,
          id: 1,
        },
      ];
      source.expose = [{ moduleId: 1, name: 'FooLib' }];
      const result = source.generate();

      const { obj } = runWithoutApi(result.content.toString());
      expect(obj['FooLib']).toEqual({ Foo: 'bar' });
    });

    it('should expose on window with target electron', () => {
      const target = 'electron';
      const source = bundleSource({ target: target });

      source.modules = [
        {
          contents: `exports.Foo = "bar"`,
          id: 1,
        },
      ];
      source.expose = [{ moduleId: 1, name: 'FooLib' }];
      const result = source.generate();

      const { obj } = runWithoutApi(result.content.toString(), { target });
      expect(obj['FooLib']).toEqual({ Foo: 'bar' });
    });

    it('should expose using "exports" for server', () => {
      const target = 'server';
      const source = bundleSource({ target: target });

      source.modules = [
        {
          contents: `exports.Foo = "bar"`,
          id: 1,
        },
      ];
      source.expose = [{ moduleId: 1, name: 'FooLib' }];
      const result = source.generate();
      const { serverExports } = runWithoutApi(result.content.toString(), { target });
      expect(serverExports['FooLib']).toEqual({ Foo: 'bar' });
    });
  });
});
