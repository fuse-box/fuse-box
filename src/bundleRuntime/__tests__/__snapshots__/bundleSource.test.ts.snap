// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundle source test Source maps should write source maps 1`] = `"{\\"version\\":3,\\"sources\\":[\\"first.js\\",\\"second/index.js\\"],\\"names\\":[\\"console\\",\\"log\\",\\"foo_1\\",\\"oi_2\\"],\\"mappings\\":\\";;;;;;;AAEAA,QAAQC,IAAIC,WAAKC;AAEjBH,QAAQC,IAAI;;AAEZD,QAAQC,IAAI;;ACJZD,QAAQC,IAAIC,WAAKC;AAEjBH,QAAQC,IAAI;;AAEZD,QAAQC,IAAI\\",\\"file\\":\\"\\",\\"sourcesContent\\":[\\"console.log(foo)\\",\\"console.log(foo)\\"]}"`;
